File: apps/worker/rubiks_worker/harmonize.py
"""Probe harmonization, in exact parity with packages/clocks/src/harmonize.ts.

1. EPICv2 replicate-suffixed probes ("cg00000029_TC21") collapse to their base
   CpG id, replicates averaged (the sesame strategy).
2. Probes dropped between array generations become candidates for reference
   imputation in the engine; harmonization never invents a value.
3. Betas outside [0,1] are evidence of malformed or non-beta input and are
   dropped (counted), not clamped.
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass
from typing import Mapping

EPIC_V2_SUFFIX = re.compile(r"_[A-Z]{2}\d+$")

METHYLATION_PLATFORMS = {"methylation_450k", "methylation_epic", "methylation_epic_v2"}


@dataclass
class HarmonizationResult:
    betas: dict[str, float]
    collapsed_replicates: int
    dropped_out_of_range: int
    dropped_non_finite: int


def harmonize_betas(platform: str, raw: Mapping[str, float]) -> HarmonizationResult:
    sums: dict[str, list[float]] = {}
    collapsed = 0
    out_of_range = 0
    non_finite = 0

    for raw_probe, value in raw.items():
        if value is None or not math.isfinite(value):
            non_finite += 1
            continue
        if value < 0 or value > 1:
            out_of_range += 1
            continue

        probe = raw_probe
        if platform == "methylation_epic_v2" and EPIC_V2_SUFFIX.search(raw_probe):
            probe = EPIC_V2_SUFFIX.sub("", raw_probe)

        bucket = sums.get(probe)
        if bucket is not None:
            bucket[0] += value
            bucket[1] += 1
            collapsed += 1
        else:
            sums[probe] = [value, 1]

    betas = {probe: total / n for probe, (total, n) in sums.items()}
    return HarmonizationResult(
        betas=betas,
        collapsed_replicates=collapsed,
        dropped_out_of_range=out_of_range,
        dropped_non_finite=non_finite,
    )

File: packages/clocks/src/harmonize.ts
import type { Platform } from "./types";

/**
 * Probe harmonization across Illumina array generations.
 *
 * The concrete problems this solves, explicitly and nowhere else:
 *
 * 1. EPICv2 renamed probes with replicate suffixes ("cg00000029_TC21").
 *    Clocks are keyed on base CpG ids, so suffixed probes are collapsed to
 *    their base id, replicates averaged (the strategy used by sesame).
 * 2. EPIC and EPICv2 dropped a subset of 450k probes that older clocks were
 *    trained on. Those become candidates for reference imputation in the
 *    engine, which records `probesImputed` and refuses above the clock's
 *    threshold. Harmonization itself never invents a value.
 * 3. Beta values outside [0,1] are evidence of a malformed or non-beta input
 *    and are dropped (counted), not clamped.
 */

export interface HarmonizationResult {
  betas: Map<string, number>;
  collapsedReplicates: number;
  droppedOutOfRange: number;
  droppedNonFinite: number;
}

const EPIC_V2_SUFFIX = /_[A-Z]{2}\d+$/;

export function harmonizeBetas(
  platform: Platform,
  raw: ReadonlyMap<string, number>,
): HarmonizationResult {
  const sums = new Map<string, { total: number; n: number }>();
  let collapsedReplicates = 0;
  let droppedOutOfRange = 0;
  let droppedNonFinite = 0;

  for (const [rawProbe, value] of raw) {
    if (!Number.isFinite(value)) {
      droppedNonFinite += 1;
      continue;
    }
    if (value < 0 || value > 1) {
      droppedOutOfRange += 1;
      continue;
    }

    let probe = rawProbe;
    if (platform === "methylation_epic_v2" && EPIC_V2_SUFFIX.test(rawProbe)) {
      probe = rawProbe.replace(EPIC_V2_SUFFIX, "");
    }

    const bucket = sums.get(probe);
    if (bucket) {
      bucket.total += value;
      bucket.n += 1;
      collapsedReplicates += 1;
    } else {
      sums.set(probe, { total: value, n: 1 });
    }
  }

  const betas = new Map<string, number>();
  for (const [probe, { total, n }] of sums) betas.set(probe, total / n);

  return { betas, collapsedReplicates, droppedOutOfRange, droppedNonFinite };
}
